GeometryService/src/PSEnclosureMaker.cc

```diff
@@ -55,62 +98,93 @@ namespace mu2e {
 
       const CLHEP::Hep3Vector shellOriginInMu2e(psEndRefPoint + CLHEP::Hep3Vector(0,0, -0.5*shellLength));
 
-      res = std::unique_ptr<PSEnclosure>(new PSEnclosure(
-			    // conical frustrum
-			    Cone(0.5*(shellODWest - 2*shellThickness),
-				 0.5*shellODWest,
-				 0.5*(shellODEast - 2*shellThickness),
-				 0.5*shellODEast,
-				 0.5*shellLength,
-				 0., CLHEP::twopi,
-				 shellOriginInMu2e,
-				 CLHEP::HepRotation(CLHEP::HepRotation::IDENTITY),
-				 shellMaterialName),
-
-			    // end plate
-			    Tube(shellMaterialName,
-				 // The vacuum volume is flush to the PS surface
-				 psEndRefPoint + CLHEP::Hep3Vector(0,0, -shellLength - 0.5*endPlateThickness),
-				 0.,
-				 0.5*shellODWest,
-				 0.5*endPlateThickness
-				 )                              
-							 )
-			    );
+                            // conical frustrum
+      Cone shellCone(0.5*(shellODWest - 2*shellThickness),
+                     0.5*shellODWest,
+                     0.5*(shellODEast - 2*shellThickness),
+                     0.5*shellODEast,
+                     0.5*shellLength,
+                     0., CLHEP::twopi,
+                     shellOriginInMu2e,
+                     CLHEP::HepRotation(CLHEP::HepRotation::IDENTITY),
+                     shellMaterialName);
+      if(vers == 2) {
+        res = std::unique_ptr<PSEnclosure>
+          (new PSEnclosure(
+                           shellCone,
+                           // end plate
+                           Tube(shellMaterialName,
+                                // The vacuum volume is flush to the PS surface
+                                psEndRefPoint + CLHEP::Hep3Vector(0,0, -shellLength - 0.5*endPlateThickness),
+                                0.,
+                                0.5*shellODWest,
+                                0.5*endPlateThickness
+                                )
+                           )
+           );
+      } else {
+        std::vector<double> zPlanes; c.getVectorDouble("PSEnclosure.endPlate.zPlanes", zPlanes);
+        std::vector<double> rIns   ; c.getVectorDouble("PSEnclosure.endPlate.rIns"   , rIns   );
+        std::vector<double> rOuts  ; c.getVectorDouble("PSEnclosure.endPlate.rOuts"  , rOuts  );
+        // double zlength = 0.;
+        // if(zPlanes.size() > 0 ) {
+        //   zlength = abs(zPlanes[0] - zPlanes[zPlanes.size() - 1]); //planes should be ordered
+        // }
+        res = std::unique_ptr<PSEnclosure>
+          (new PSEnclosure(shellCone,
```

Comment:
Suggest

res = std::make_unique<PSEnclosure>(shellCone, ...);

Reply:
I'm not sure how to make this change, simply replacing `std::unique_ptr` with `std::make_unique` at this line for example doesn't compile. Do I have to make additional changes elsewhere?

Reply:
Here's an example of a change I tried to make:
```
        res = std::unique_ptr<PSEnclosure>
          (new PSEnclosure(
                           shellCone,
                           // end plate
                           Tube(shellMaterialName,
                                // The vacuum volume is flush to the PS surface
                                psEndRefPoint + CLHEP::Hep3Vector(0,0, -shellLength - 0.5*endPlateThickness),
                                0.,
                                0.5*shellODWest,
                                0.5*endPlateThickness
                                )
                           )
           );
```
converted to:
```
        res = std::make_unique<PSEnclosure>
          (
           shellCone,
           // end plate
           Tube(shellMaterialName,
                // The vacuum volume is flush to the PS surface
                psEndRefPoint + CLHEP::Hep3Vector(0,0, -shellLength - 0.5*endPlateThickness),
                0.,
                0.5*shellODWest,
                0.5*endPlateThickness
                )
           );

```
runs into this error:
```
/cvmfs/mu2e.opensciencegrid.org/artexternals/gcc/v9_3_0/Linux64bit+3.10-2.17/include/c++/9.3.0/bits/unique_ptr.h: In instantiation of 'typename std::_MakeUniq<_Tp>::__single_object std::make_unique(_Args&& ...) [with _Tp = mu2e::PSEnclosure; _Args = {mu2e::Cone&, mu2e::Tube}; typename std::_MakeUniq<_Tp>::__single_object = std::unique_ptr<mu2e::PSEnclosure>]':
Offline/GeometryService/src/PSEnclosureMaker.cc:123:12:   required from here
/cvmfs/mu2e.opensciencegrid.org/artexternals/gcc/v9_3_0/Linux64bit+3.10-2.17/include/c++/9.3.0/bits/unique_ptr.h:857:30: error: 'mu2e::PSEnclosure::PSEnclosure(const mu2e::Cone&, const mu2e::Tube&, int)' is private within this context
  857 |     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from Offline/GeometryService/src/PSEnclosureMaker.cc:10:
./Offline/ProductionSolenoidGeom/inc/PSEnclosure.hh:70:5: note: declared private here
   70 |     PSEnclosure (const Cone& shellCone, const Tube& ep, int vers=2 )
      |     ^~~~~~~~~~~
```

Reply:
Well, if the constructor was made public in PSEnclosure.hh, this particular line should be happy.
I don't know if this breaks anything else, though.   I've not done a lot with make_unique.

Reply:
I agree, this seems to be the issue. I suppose the question is then is it better to make the constructor public and use make_unique, or leave it private and use unique_ptr? Perhaps Andrei has a preference?

Reply:
For now just make the c'tor public.  Whoever created this did not follow the pattern.  The design is that a Geometry object is meant to be a simple container of information that does not know how to construct itself.  The matching maker class was meant to have all of the smarts about how to build it.  The idea is that the maker is allowed to depend on arbitrarily complex external things, like conditions, databases, other parts of the but the geometry object class should be simple.  If you do this right then the users of the geometry object do not have any compile time or run time dependence on the infrastructure that knows how to make the geometry object.  

Part of that pattern was the at the c'tor of the geometry object was private and friended to the maker.  

For now just make it public.

Reply:
Strange.   There is no requirement that the template arg of make_unique is default constructable.   Nothing in the c++20 standard draft, nothing  in my internet searches.   And a simple standalone test happily builds a test case below,  make_unique without default ctr.

The reason I would like to avoid default constructors is the same as not having uninitialized variables in general.
As a silly example,  it would allow one to compile something like

PSEnclosure pe;
function_that_needs_enclosure_geometry(pe);

without remembering to use the geometry service.
And what is the benefit of having default constructor for such an object?

Here is the test case

//================================================================
// compile with
// g++ -std=c++17

#include <memory>
#include <iostream>

class C {
  int x_;
  int y_;
  C() {}
public:
  C(int x,int y) : x_{x}, y_{y} {}
  int sum() { return x_ + y_; }
};

int main() {
  auto res = std::make_unique<C>(1,2);
  std::cout<<"OK"<<std::endl;
  return 0;
}
//================================================================

Reply:
Ah, not the default constructor but all the normal constructors.   Yes, that makes sense.   I just tried with the actual Offline code, and the default constructor can be kept private.  If the real constructors are public then everything compiles.

The (private) default constructor declaration can probably be removed from the code; the reason I added it in the first place was to do I/O of geometry objects into art files, but Rob discourages that.

Reply:
One the reason behind the design is exactly the case that Andrei illustrated. Has anyone looked to see how widely this pattern has been broken?

Reply:
I made the non-default constructors public and switched to `make_unique`. I tried removing the default constructor and ran into some issues doing this. Perhaps this is best done on a separate PR?

Reply:
Sure, please leave the private default constructor alone if its removal causes any problems.